### PR TITLE
SONARJAVA-1615 Filter issues related to Eclipse I18N framework

### PR DESCRIPTION
**Pull request by m-g-sonar**

### CODE REVIEW

File: java-frontend/src/main/java/org/sonar/java/SonarComponents.java [hunk 1]
```diff
@@ -139,12 +144,21 @@ public ResourcePerspectives getResourcePerspectives() {
...
+  }
+
   public void registerCheckClasses(String repositoryKey, List<Class<? extends JavaCheck>> checkClasses) {
     checks.add(checkFactory.<JavaCheck>create(repositoryKey).addAnnotatedChecks(checkClasses));
   }
 
   public CodeVisitor[] checkClasses() {
     List<CodeVisitor> visitors = Lists.newArrayList();
+    // add issue filters BEFORE all the other checks
+    for (JavaIssueFilter issueFilter : issueFilters) {
```
Comment (benzonico) [inline java-frontend/src/main/java/org/sonar/java/SonarComponents.java]: addAll ?

File: java-frontend/src/main/java/org/sonar/java/filters/BaseTreeVisitorIssueFilter.java [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+
+  @Override
+  public boolean accept(Issue issue) {
+    if (targetedRules().contains(issue.ruleKey().rule()) && isIgnoredLine(issue.componentKey(), issue.line())) {
+      return false;
+    }
+    return true;
+  }
+
+  public abstract Set<String> targetedRules();
```
Comment (benzonico) [inline java-frontend/src/main/java/org/sonar/java/filters/BaseTreeVisitorIssueFilter.java]: Name is debatable (filteredRules ? ) but moreover : as it is implemented right now you will build a new ImmutableSet for EVERY issue tested.

File: java-frontend/src/main/java/org/sonar/java/filters/PostAnalysisIssueFilters.java [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+
+import com.google.common.collect.ImmutableList;
+
+import org.sonar.api.issue.Issue;
+import org.sonar.api.issue.batch.IssueFilter;
+import org.sonar.api.issue.batch.IssueFilterChain;
+
+public class PostAnalysisIssueFilters implements IssueFilter {
+
+  private final Iterable<JavaIssueFilter> issueFitlers;
```
Comment (benzonico) [inline java-frontend/src/main/java/org/sonar/java/filters/PostAnalysisIssueFilters.java]: typo !

File: java-frontend/src/main/java/org/sonar/java/filters/BaseTreeVisitorIssueFilter.java (hunk 2 above)
Comment (benzonico) [inline java-frontend/src/main/java/org/sonar/java/filters/BaseTreeVisitorIssueFilter.java]: I would rather construct the key Set in the constructor of this object.

File: java-frontend/src/main/java/org/sonar/java/DefaultJavaResourceLocator.java [hunk 4]
```diff
@@ -126,19 +130,29 @@ public void scanFile(JavaFileScannerContext context) {
...
       }
     }
     methodStartLines.putAll(javaFilesCache.getMethodStartLines());
+    String componentKey = currentResource.getEffectiveKey();
     if (javaFilesCache.hasSuppressWarningLines()) {
-      suppressWarningsFilter.addComponent(currentResource.getEffectiveKey(), javaFilesCache.getSuppressWarningLines());
+      suppressWarningsFilter.addComponent(componentKey, javaFilesCache.getSuppressWarningLines());
+    }
+    for (JavaIssueFilter javaIssueFilter : issueFilters) {
+      javaIssueFilter.addComponent(fileKey, componentKey);
```
Comment (benzonico) [inline java-frontend/src/main/java/org/sonar/java/DefaultJavaResourceLocator.java]: For what I can read you only need to set the component key, no need for an internal map on filters objects.

Comment (benzonico) [inline java-frontend/src/main/java/org/sonar/java/DefaultJavaResourceLocator.java]: Therefore you should have a setComponentKey method only.

File: java-frontend/src/main/java/org/sonar/java/filters/EclipseI18NFilter.java [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+import com.google.common.collect.ImmutableSet;
+
+import org.sonar.plugins.java.api.tree.ClassTree;
+
+import java.util.Set;
+
+public class EclipseI18NFilter extends BaseTreeVisitorIssueFilter {
+
+  @Override
+  public Set<String> targetedRules() {
```
Comment (benzonico) [inline java-frontend/src/main/java/org/sonar/java/filters/EclipseI18NFilter.java]: I would really rely on class name and use the AnnotationUtils class : 
```
org.sonar.check.Rule ruleAnnotation = AnnotationUtils.getAnnotation(clazz, org.sonar.check.Rule.class);
``` 
to retrieve the key and contruct this set of string. 

This way, we prevent ourselves of key change that are bound to happen and there is only one place where the key is defined.

File: java-frontend/src/main/java/org/sonar/java/filters/EclipseI18NFilter.java [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+    return ImmutableSet.of(
+      // "public static" fields should be constant
+      "S1444",
+      // Class variable fields should not have public accessibility (with legacy key)
+      "S1104", "ClassVariableVisibilityCheck");
```
Comment (benzonico) [inline java-frontend/src/main/java/org/sonar/java/filters/EclipseI18NFilter.java]: Why legacy and new key ? there is only one rule key !

File: java-frontend/src/main/java/org/sonar/java/filters/BaseTreeVisitorIssueFilter.java [hunk 7]
```diff
@@ -0,0 +1,105 @@
...
+
+import javax.annotation.Nullable;
+
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Set;
+
+public abstract class BaseTreeVisitorIssueFilter extends BaseTreeVisitor implements JavaIssueFilter {
+
+  @Nullable
```
Comment (benzonico) [inline java-frontend/src/main/java/org/sonar/java/filters/BaseTreeVisitorIssueFilter.java]: Not nullable.

File: java-frontend/src/main/java/org/sonar/java/filters/BaseTreeVisitorIssueFilter.java [hunk 8]
```diff
@@ -0,0 +1,105 @@
...
+  private String componentKey;
+  private final Map<String, Set<Integer>> ignoredLinesByComponent = Maps.newHashMap();
```
Comment (benzonico) [inline java-frontend/src/main/java/org/sonar/java/filters/BaseTreeVisitorIssueFilter.java]: MultiMap

File: java-checks/src/main/java/org/sonar/java/filters/PostAnalysisIssueFilter.java [hunk 9]
```diff
@@ -0,0 +1,70 @@
...
+        return false;
+      }
+    }
+    return chain.accept(issue);
+  }
+
+  @Override
+  public void scanFile(JavaFileScannerContext context) {
+    InputFile inputFile = fs.inputFile(fs.predicates().is(context.getFile()));
+    org.sonar.api.resources.File currentResource = (org.sonar.api.resources.File) sensorContext.getResource(inputFile);
```
Comment (SonarTech) [inline java-checks/src/main/java/org/sonar/java/filters/PostAnalysisIssueFilter.java]: MAJOR org.sonar.java.filters.PostAnalysisIssueFilter must not use org.sonar.api.resources.File [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3ADeprecated_org_sonar_api_resources_File_should_not_be_used)

**Comment by SonarTech** [comment]
SonarQube analysis reported 1 issue
* MAJOR 1 major

Watch the comments in this conversation to review them.

File: java-checks/src/main/java/org/sonar/java/filters/PostAnalysisIssueFilter.java (hunk 9 above)
Comment (m-g-sonar) [inline java-checks/src/main/java/org/sonar/java/filters/PostAnalysisIssueFilter.java]: I would love to rely only on the InputFile, but the Issue interface only provides a component key to identify the file.

File: java-frontend/src/main/java/org/sonar/java/DefaultJavaResourceLocator.java [hunk 10]
```diff
@@ -24,6 +24,7 @@
 import com.google.common.collect.ImmutableList;
 import com.google.common.collect.ImmutableSortedSet;
 import com.google.common.collect.Maps;
+
```
Comment (benzonico) [inline java-frontend/src/main/java/org/sonar/java/DefaultJavaResourceLocator.java]: Those modification are not anymore relevant to your fix.

File: java-frontend/src/test/java/org/sonar/java/DefaultJavaResourceLocatorTest.java [hunk 11]
```diff
@@ -20,6 +20,7 @@
 package org.sonar.java;
 
 import com.google.common.collect.Lists;
+
```
Comment (benzonico) [inline java-frontend/src/test/java/org/sonar/java/DefaultJavaResourceLocatorTest.java]: useless modification

File: java-frontend/src/test/java/org/sonar/java/JavaVersionAwareVisitorTest.java [hunk 12]
```diff
@@ -22,6 +22,7 @@
 import com.google.common.base.Charsets;
 import com.google.common.collect.ImmutableList;
 import com.google.common.collect.Lists;
+
```
Comment (benzonico) [inline java-frontend/src/test/java/org/sonar/java/JavaVersionAwareVisitorTest.java]: useless modification

**Pull request merged**